src/ui/actions/timeline.ts

```diff
@@ -32,14 +37,79 @@ export type TimelineActions =
   | SetZoomRegionAction
   | SetHoveredPoint;
 
+const GET_FIRST_COMMENT_POINT = gql`
+  query GetFirstCommentTime($recordingId: uuid) {
+    comments(
+      where: { recording_id: { _eq: $recordingId }, _and: { parent_id: { _is_null: true } } }
+      order_by: { time: asc }
+      limit: 1
+    ) {
+      time
+      point
+      has_frames
+    }
+  }
+`;
+
 export async function setupTimeline(recordingId: RecordingId, store: UIStore) {
-  const { dispatch } = store;
+  const { dispatch, getState } = store;
   ThreadFront.on("paused", args => dispatch(onPaused(args)));
-  ThreadFront.on("endpoint", args => dispatch(onEndpoint(args)));
   ThreadFront.warpCallback = onWarp(store);
   const description = await ThreadFront.getRecordingDescription();
   dispatch(setRecordingDescription(description));
   window.addEventListener("resize", () => dispatch(updateTimelineDimensions()));
+
+  await ThreadFront.waitForSession();
+  const { endpoint } = await client.Session.getEndpoint({}, ThreadFront.sessionId!);
+  const { point, time } = endpoint;
+  if ("lastScreen" in description && description.lastScreen) {
+    addLastScreen(description.lastScreen, point, time);
+  }
+
+  const zoomRegion = selectors.getZoomRegion(getState());
+  const newZoomRegion = { ...zoomRegion, endTime: time };
+  dispatch(
+    setTimelineState({ currentTime: time, recordingDuration: time, zoomRegion: newZoomRegion })
+  );
+
+  if (!getTest()) {
+    const firstCommentResult = await query({
+      query: GET_FIRST_COMMENT_POINT,
+      variables: { recordingId },
+    });
+    const firstComment = firstCommentResult?.data?.comments?.[0];
+    if (firstComment) {
+      const { point, time, has_frames } = firstComment;
+      ThreadFront.timeWarp(point, time, has_frames);
+      return;
+    }
+
+    await paintPointsWaiter;
+    for (const paintPoint of gPaintPoints.slice(0, 10)) {
```

Comment:
this feels very soupy...

Reply:
i think graphics should hide most of this from redux... also others who import the protocol might want it.

Reply:
I like the cleanup you did there.